e3/os/process.py

```diff
@@ -625,29 +625,49 @@ def wait_for_processes(process_list, timeout):
                 raise WaitError
 
     else:
-        import e3.os.unix.process
-        # Choose between blocking or non-blocking call to wait
-        blocking = True
-        if timeout != 0:
-            blocking = False
-
-        while remain >= 0.0 or timeout == 0:
-            # Retrieve first child process that ends. Note that that child
-            # process might not be in our watch list
-            pid = e3.os.unix.process.wait(blocking)
-
-            if pid != 0:
-                # We have a result
-                result = next((index for index, p in
-                               enumerate(process_list) if p.pid == pid), None)
-                if result is not None:
-                    # At the stage we need to set the process status and close
-                    # related handles. Indeed we will not be able to use the
-                    # wait method afterwards and retrieve it.
-                    process_list[result].wait()
-                    return result
-            time.sleep(1.0)
-            remain = timeout - time.time() + start
+        import select
+        import signal
+
+        # Each time a SIGCHLD signal is received write into pipe. Use
+        # then select which support timeout arguments to wait.
+        fd_r, fd_w = os.pipe()
+
+        def handler(signum, frame):
+            os.write(fd_w, b'a')
+
+        signal.signal(signal.SIGCHLD, handler)
+
+        try:
+            while remain >= 0.0 or timeout == 0:
+                # Do a first check in case a SIGCHLD was emited before the
+                # initialisation of the handler.
+                for index, p in enumerate(process_list):
+                    if p.poll() is not None:
+                        return index
+
+                # Wait for a sigchld signal. Note that select might
+                # be interrupted by signals thus the loop
+                while True:
+                    select_args = [[fd_r], [], []]
```

Comment:
do we need to keep select_args in the loop? can we move that before `while remain >=...`?